### PR TITLE
Refactor S3 model logic into abstract base classes

### PR DESCRIPTION
**Pull request by Spacerat**
This is an enabling PR for the test-bench branch, but is also an improvement in its own right. It makes the following changes:

- All S3 logic for saving/reading of files attached to SourceFile and ResourceVariant models is factored out into a single S3File class which retains all previous functionality.
- Before there was  `SourceFiles.save_file(text)` and `ResourceVariants.save_file(stream)`. Now, `S3File.save_text(text)` replaces the SourceFile's version, and all usage is renamed accordingly.
- Support for line folding and last-modified behaviour is factored out of SourceFile and into ScriptFile. This is purely an enabling change for test-bench, where TestFile will inherit from ScriptFile.
- All support previously in `files.py` for `AWS_ENABLED==False` is now contained in a small set of functions in S3File, which will make them easy to remove if we ever want to.

### CODE REVIEW

File: ide/models/scriptfile.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import datetime
+
+from django.utils.timezone import now
+from django.db import models
+
+from ide.models.s3file import S3File
+
+
+class ScriptFile(S3File):
```
Comment (Katharine) [inline ide/models/scriptfile.py]: This is a bizarre name.

Comment (Spacerat) [inline ide/models/scriptfile.py]: The motivation is that it is the superclass of both SourceFile and TestFile in the test bench branch. I don't mind changing the name if you can think of something more suitable

Comment (Spacerat) [inline ide/models/scriptfile.py]: In fact, the comment is wrong. It should say "ScriptFiles add support to S3Files for last-modified timestamps and code folding". Which makes me think that TextFile is a reasonable alternative name

Comment (Katharine) [inline ide/models/scriptfile.py]: Works for me.

**Comment by Spacerat** [comment]
OK, I renamed ScriptFile to TextFile

**Pull request merged**